src/assets.js

```diff
@@ -334,22 +339,42 @@ class Assets extends Events {
      * @private
      * @param {AssetUploadArguments} data - The asset fields
      * @param {TextureImportSettings|SceneImportSettings} settings - Import settings
+     * @param {Function} onProgress - Function to report progress
      * @returns {Promise<Asset>} The new asset
      */
-    async upload(data, settings = {}) {
+    async upload(data, settings = {}, onProgress = null) {
         if (data.folder) {
             data.folderId = data.folder.get('id');
         }
 
-        const result = await uploadFile(data, settings);
-        let asset = this.get(result.id);
-        if (!asset) {
-            asset = await new Promise(resolve => {
-                this.once(`add[${result.id}]`, a => resolve(a));
-            });
+        const uploadId = this._uploadId++;
+        if (!onProgress && this._defaultUploadProgressCallback) {
+            onProgress = (progress) => {
+                this._defaultUploadProgressCallback(uploadId, progress);
+            };
         }
 
-        return asset;
+        try {
+            const result = await uploadFile(data, settings, onProgress || this._defaultUploadProgressCallback);
+            let asset = this.get(result.id);
+            if (!asset) {
+                asset = await new Promise(resolve => {
+                    this.once(`add[${result.id}]`, a => resolve(a));
+                });
+            }
+
+            if (this._defaultUploadCompletedCallback) {
+                this._defaultUploadCompletedCallback(uploadId, asset);
```

Comment:
Are we guaranteed to have an asset at this point? Is it possible for the promise on L361 to return a null asset?

Reply:
We are